implementations/elixir/ockam/ockam/lib/ockam/worker/authorization.ex

```diff
@@ -0,0 +1,129 @@
+defmodule Ockam.Worker.Authorization do
+  @moduledoc """
+  Helper functions to configure worker message authorization.
+
+  Usage:
+  ```
+  def is_authorized(message, _state)
+    Authorization.from_addresses(message, ["one", "two"])
+  end
+  ```
+
+  Pipelining helpers:
+  ```
+  def is_authorized(message, state)
+    Authorization.from_addresses(message, ["one", "two"])
+    |> Authorization.to_my_address(message, state)
+  end
+  ```
+  """
+  alias Ockam.Message
+
+  @doc """
+  Allow any messages to be handled by the worker
+  """
+  def allow_all(prev \\ :ok) do
+    chain(prev, fn -> :ok end)
+  end
+
+  @doc """
+  Deny all messages from being handled by the worker
+  """
+  def deny_all(prev \\ :ok) do
+    chain(prev, fn -> {:error, :deny_all} end)
+  end
+
+  @doc """
+  Only allow messages from a set of addresses to be handled by the worker.
+
+  Message address is taken from return_route in that case
+  """
+  def from_addresses(prev, message, addresses) when is_list(addresses) do
+    chain(prev, fn ->
+      from_addresses(message, addresses)
+    end)
+  end
+
+  def from_addresses(message, addresses) do
+    case Message.return_route(message) do
+      [return_address | _rest] ->
+        case Enum.member?(addresses, return_address) do
+          true -> :ok
+          false -> {:error, {:from_address, :invalid_source, return_address}}
+        end
+
+      [] ->
+        {:error, {:from_address, :empty_return_route}}
+    end
+  end
+
+  @doc """
+  Only allow messages sent to a certain addresses to be handled by the worker
+
+  Message address is taken from the message onward_route
+  """
+  def to_addresses(prev, message, addresses) when is_list(addresses) do
+    chain(prev, fn ->
+      to_addresses(message, addresses)
+    end)
+  end
+
+  def to_addresses(message, addresses) do
+    case Message.onward_route(message) do
+      [onward_address | _rest] ->
```

Comment:
would it make sense to enforce that this is the _final_ step in the route?

Reply:
The intention for this verification was that the message is sent to this worker address. And messages get sent to the first address in the onward route. What happens to the rest of the route is up to the worker to decide, whether it forwards it or not.
We might add another authorization rule that the worker address is the only one left in the onward route, but I'm not sure we need that right now. If we have workers implementing that we can generalize that as a rule later.